### ISSUE TITLE
Migrate to local text embeddings (replace OpenAI embeddings with local models)

### ISSUE THREAD

**Issue by britt** [britt/llpm#129]
Proposal: switch embeddings to use bge-base-en-v1.5 (Hugging Face Transformers)

Summary

We should update the existing embeddings provider to support running bge-base-en-v1.5 using Hugging Face Transformers (local hosting). This will let us run higher-quality embeddings locally (or on our own servers) and provide a drop-in fallback to OpenAI/remote providers.

Motivation

- Allow running bge-base-en-v1.5 locally (where model weights are available) to reduce latency and cost.
- Provide an easy-to-run local service that other parts of llpm can call (HTTP API) so we don't have to embed Python model runtime in the node process.
- Keep a fallback to current OpenAI provider to preserve reliability and quick rollback.

High-level plan

1) Add a small Python embedding service (FastAPI) that loads bge-base-en-v1.5 with Transformers and exposes a /embeddings endpoint that accepts a JSON array of strings and returns embeddings.
   - Use AutoTokenizer + AutoModel from transformers.
   - Use mean pooling over last_hidden_state (or the model-specific embedding extraction if documented).
   - Normalize vectors (L2) before returning.
   - Support device selection via env var (CUDA/MPS/CPU).
   - Batch inputs and support a configurable max batch size.

2) Add a TypeScript EmbeddingsProvider implementation that calls the local service (configurable URL).
   - Provide retries/backoff and a fallback to the OpenAI provider.
   - Add a feature flag or config entry to switch between providers at runtime.

3) Add a Dockerfile for the service so it’s easy to run in CI or dev.
4) Add unit tests for the provider and an integration test that starts the service and validates embeddings shape and deterministic behavior on sample inputs.
5) Add a short benchmark script in tools/ to measure latency and throughput vs OpenAI embeddings (record CPU/RAM and batch sizes).
6) Update docs and add migration notes in the README/issue.

Example Python service (FastAPI)

```python
from typing import List
from fastapi import FastAPI
from pydantic import BaseModel
import torch
from transformers import AutoTokenizer, AutoModel
import torch.nn.functional as F

MODEL_NAME = "bge-base-en-v1.5"
DEVICE = "cuda" if torch.cuda.is_available() else ("mps" if torch.backends.mps.is_available() else "cpu")

app = FastAPI()

class EmbeddingRequest(BaseModel):
    input: List[str]

class EmbeddingResponse(BaseModel):
    embeddings: List[List[float]]

# load once on startup
tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
model = AutoModel.from_pretrained(MODEL_NAME)
model.to(DEVICE)
model.eval()

@app.post("/embeddings", response_model=EmbeddingResponse)
async def embeddings(req: EmbeddingRequest):
    texts = req.input
    batch_size = 32
    all_embs = []
    with torch.no_grad():
        for i in range(0, len(texts), batch_size):
            batch = texts[i:i+batch_size]
            inputs = tokenizer(batch, padding=True, truncation=True, return_tensors="pt")
            inputs = {k: v.to(DEVICE) for k, v in inputs.items()}
            outputs = model(**inputs, return_dict=True)
            # mean pooling over sequence length
            last_hidden = outputs.last_hidden_state  # (B, T, D)
            mask = inputs.get("attention_mask")
            if mask is not None:
                mask = mask.unsqueeze(-1).expand(last_hidden.size()).float()
                summed = (last_hidden * mask).sum(dim=1)
                counts = mask.sum(dim=1).clamp(min=1e-9)
                emb = summed / counts
            else:
                emb = last_hidden.mean(dim=1)
            emb = F.normalize(emb, p=2, dim=1)
            all_embs.extend(emb.cpu().tolist())
    return {"embeddings": all_embs}
```

Device & performance notes

- On macOS with Apple Silicon, set DEVICE to mps for best performance with PyTorch MPS builds.
- For larger models or GPU servers, use CUDA.
- Batch size tuning: start with 16–64 depending on memory.
- Consider using ONNX or gguf/ggml conversions for even faster inference if needed.

Licensing

- Confirm bge-base-en-v1.5 weights are available on Hugging Face and that license permits local hosting. If weights are not redistributable, we cannot bundle them in Docker images.

Testing & rollout

- Add unit tests for the provider (mock service) and an integration test that runs the FastAPI service in CI (or use a local test mode that runs the model in-process).
- Add a feature flag / ENV var to enable the HF provider and keep OpenAI as the default until validated.
- Run benchmarks on CI or a Mac dev machine and document latency/throughput.

Next steps (I can do any of these):
- Implement the FastAPI service + Dockerfile and add to tools/ or infra/.
- Implement the TypeScript EmbeddingsProvider that calls the service and add a config toggle.
- Open a PR with the changes and tests.

If you want me to proceed, tell me which next step (implement service, implement TS provider, or open PR with both). Also confirm model availability & license for bge-base-en-v1.5 if you have that info.
